### ISSUE TITLE
[ENH] using window summary features in reduction strategies (with scikit-learn regressors)

### ISSUE THREAD

**Issue by Dwarak-PCI** [alan-turing-institute/sktime#1577]
**Describe the solution you'd like**
I would to forecast time series data using scikit learn regressors with moving average smoothing. Currently, sktime module supports lagged time series component (window_length) which is specified in make_reduction method. It would be of great value, if this method supports moving average smoothing (backward rolling window) for target time series. 

**Describe alternatives you've considered**
Alternative solution is to use ARIMA, because it supports both Auto Regressive component (lags) and Moving Average component (Smoothing). But it is not as good as GBT algorithm.

**Additional context**
I understand that tsfresh provides tools to extract features but this process takes a lot more time to compute the features.

**Comment by fkiraly** [comment]
Interesting - I was wondering whether there is already a way to do this.

The generic way would be something like a `SmoothingReducer(base_estimator, smoother_estimator)`.

This does the following in `predict`:

1. forecast time points `1...max(fh)` using `base _estimator`, this gives a series `x(0) ... x(t) ... x(t+maxfh)` made up of actuals and forecasts
2. using `smoother_estimator`, produce forecasts `z(i)` for all `i` in `fh` from `x(0),...,x(i-1)` using `update_predict`, always one-step ahead
3. the forecast are the `z(i)` for `i` in `fh`

In `fit`, the `smoother_estimator` is trained on `y` (observed) to make one-stepahead forecasts.

**Comment by aiwalter** [comment]
@danbartl

**Comment by Dwarak-PCI** [comment]
@fkiraly 
Thanks for the quick response. I couldn't find SmoothingReducer method in the API documentation. How can I import this method?
```

regressor = LGBMRegressor(n_estimators=200, random_state=0)
#regressor = Lasso(copy_X=True)
my_reduced_regressor_estimator = make_reduction(
    regressor,
    strategy="recursive",
    window_length=24,
    scitype="tabular-regressor",
)
model = SmoothingReducer(my_reduced_regressor_estimator, ARMA(p=0, q=4))
fh = np.arange(1, parameters["forecast_period"] + 1)
model.fit(y_train.copy(), x_train.copy())
predictions = model.predict(
    fh, X=x_test.copy()
)
```

Pandas rolling function does this type of rolling average:
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rolling.html

Pandas rolling method allows the user to enter window_length, weather we want to average, min, max etc. I am currently using this method on exogenous data to compute rolling window min/max/avg, but I cannot do this with target data (y). I guess, this (rolling window) needs to be handled the same way window_length (lags) are being implemented currently.

**Comment by Dwarak-PCI** [comment]
That is correct, I would want something similar to `SmoothingReducer(my_reduced_regressor_estimator, ARMA(p=0, q=my_q))`

If p > 0, then it works similar to lagged regressors. How would this be different from window_length provided in make_reduction method?

**Comment by fkiraly** [comment]
>How would this be different from window_length provided in make_reduction method?

I think that's simply a different parameter? The `window_length` in `make_reduction` affects the lags that go into `base_estimator`, not the `smoothing_estimator`.

Do you mean, you think there is a way how one could achieve the same algorithm using only `make_reduction`? I do not see this, but that doesn't mean it's wrong - can you elaborate?

**Comment by Dwarak-PCI** [comment]
Sure.

Lets consider lags first:
Given that y is our target_variable, our goal is to predict y(t).
if window_length = 4, which means I am going to have y(t-1), y(t-2), y(t-3), y(t-4) as exogenous data to predict y(t). Do you agree with this?

Now, lets switch to moving average:
Given that y is our target_variable, our goal is the predict the same y(t).
If moving_average_window_width = 4, then my exogenous variable x(t) can be calculated as follows:
```
x(t) = Average[y(t-1), y(t-2), y(t-3), y(t-4)]
or 
x(t) = Min[y(t-1), y(t-2), y(t-3), y(t-4)]
or
x(t) = Max[y(t-1), y(t-2), y(t-3), y(t-4)]
```

x(t) needs to be computed at every forecast time step, just like lags. Does this example make sense?
Please let me know if you would like review this over a zoom/discord call?
 
Ideally, I would like the following parameters to be added to make_reduction method:

`my_reduced_regressor_estimator = make_reduction(
regressor,
strategy="recursive",
window_length=24,
**rolling_window_width = 4,
rolling_window_statistics = ["mean","max","min"],**
scitype="tabular-regressor",
)`

Just like how lags goes into base_estimator, is it possible to make base_estimator accept rolling_window field?

**Comment by RNKuhns** [comment]
I had been thinking that we should a set of smoothing transformers (they should do simple smoothing, but we could also have a spline one and other smoothing/filtering methods).

Since this is also applicable in the non-reduction case (you might smooth/filter a series then forecast it; I believe there is even a paper I’ve read that showed this sometimes improves accuracy).

Then use that transformer in a regular pipeline.

I am also curious how does make_reduction currently interest with pipelines?

**Comment by LoicRaillon** [comment]
As mentioned by @Dwarak-PCI it would be nice to have the possibility to use a list of aggregate functions (mean, median, min, etc.) on the lagged windows of the target variable `y` or the exogenous variables `X` and then make a recursive forecast as `make_reduction` does.

An interesting example is load curve forecasting where the outdoor temperature has a delayed influence on the target. One way to model this is to use a low-pass filter on the outdoor temperature, where the time constant of the low-pass filter (which is an estimate of the building heat loss coefficient) is an hyper-parameter which can be estimated with cross-validation. 

I think it  could be achieved by using a mapping between variable names and aggregate functions:
`lags = {"y": [1, 2, 3], "X1": [3]}`
`aggregates = {"y": {"window": 4, "funs": [mean]}, "X2": {"window": 8, "funs": [median, skewness]}`

**Comment by Dwarak-PCI** [comment]
@LoicRaillon Thanks for summarizing my thoughts. This is exactly what I am looking for. Temperature have a lagged effect on load curve, and also in case of wind power forecasting (wind farm), wind speed tend to have a lagged effect on the wind power. To address this, I can create lags and aggregates on wind speed or outdoor temperature. But it is not possible for me to create aggregate on target series (load curve or wind power) that I am trying to forecast.

**Comment by fkiraly** [comment]
> x(t) needs to be computed at every forecast time step, just like lags. Does this example make sense?

Ahhh, @Dwarak-PCI, I think I get what you want.
You want exogeneous variables in the reduction that can be created by some aggregation or transformer operation.
Then apply the current `make_reduction`, using these summaries as features.

We have already identified the "window aggregation" operation as a thing we need - @danbartl was working on this, although there seems to be no issue that he opened. I've created this in #1581 now.

Once that is implemented, you can get what you want as follows, as long as your reduction strategy already supports use of exogeneous variables:

```python
my_lag_aggregator = WindowSummarizer(lags, aggregates) # depending on the signature that this will have

my_reduced_regressor_estimator = make_reduction(
    regressor,
    strategy="recursive",
    window_length=24,
    scitype="tabular-regressor",
)

my_pipeline = ForecastingPipeline(steps=[
    ("lag_aggregator", mylag_aggregator),
    ("reducer", my_reduced_regressor_estimator)
])
```

**Comment by fkiraly** [comment]
FYI @danbartl, @ltsaprounis

**Comment by fkiraly** [comment]
Question to @Dwarak-PCI, @LoicRaillon, am I correctly understanding *now* what you want?

If so, would it be solved by the `WindowSummarizer` (or whatever it ends up being called) that @danbartl is implementing?

**Comment by fkiraly** [comment]
> Please let me know if you would like review this over a zoom/discord call?

Feel free to join our regular collab sessions on Fridays, 1pm UTC - might be worth pining @danbartl and @ltsaprounis in advance and/or meet, since it seems very much related.

**Comment by Dwarak-PCI** [comment]
@fkiraly 
Yes, WindowSummarizer is exactly what I was looking for. I have a question for you:

WindowSummarizer(lags, aggregates)
Can aggregates take min, max, sum, std, mean, median etc as values?
Thanks for your time.

**Comment by LoicRaillon** [comment]
@fkiraly `WindowSummarizer` should work for exogenous variables. It might be interesting to have the same capabilities for the target variable `y` in `make_reduction`.

**Comment by fkiraly** [comment]
@LoicRaillon, if it´s a transformer, you can use it on `y` via `TransformedTargetForecaster` or on `X` via `ForecastingPipeline`.

What would be the case to attach it to `make_reduction` instead of having it as a separate transformer? Windowing/summarization seems to be a separate operation from the reduction step to me.

**Comment by LoicRaillon** [comment]
@fkiraly Maybe I miss something but if we use a moving average for the target `y` on window of `N` lags, e.g. `X[t] = (y[t], y[t-1], ..., y[t-N-1]) / N`, we can forecast `y_hat[t+1] = f(X[t])`. In order to forecast `y_hat[t+2] = f(X[t+1]` we need to append `y_hat[t+1]` to the window of lags, e.g. `X[t+1] = (y_hat[t+1], y[t], y[t-1], ..., y[t-N-2]) / N`. This recursion is done in `make_reduction` but only for simple lags and not for aggregate functions of lags. Does it make sense or I don't see the obvious?

**Comment by fkiraly** [comment]
Ah, I see - you want to use the summaries of `y` to forecast `y`, and this is not possible with the current pipeline constructs, even if `WindowSummarizer` were available, is that what you mean?

Because for that you would need a pipeline that puts computes summaries of `y` but treats them as `X` in `make_reduction`?

**Comment by fkiraly** [comment]
@aiwalter, quick quesiton: is there currently a way with the pipelines to compute transformations of `y` and "consider them as `X`"?

**Comment by aiwalter** [comment]
@fkiraly I think what you describe is here in the table as case 4: https://github.com/sktime/enhancement-proposals/blob/main/steps/01_forecasting_api/forecasting_pipeline.md

We could possibly cover it with a `Featurizer` as described in my issue #1443 and an additional param `swap_y_X` or sth?

So this would be 
```python
pipe = ForecastingPipeline(steps=[
    ("boxcox", Featurizer(BoxCox(), swap_y_X=True)),
    ("forecaster", NaiveForecaster()),
]
pipe.fit(Z=X, X=y)

# then we internally do this in Featurizer:
bc = BoxCox()
bc.fit_transform(Z=y, X=X)
# ... and append the output to X as new column
```

**Comment by ltsaprounis** [comment]
I was thinking about this also in the context of the recursive case. I don't see how we can implement this as a transformer and then use it in a pipeline. In my opinion we need to add this as an extra feature in make reduction.

```python
my_reduced_regressor_estimator = make_reduction(
    regressor,
    strategy="direct",
    window_length=24,
    summary_features={"mean": (lags, rolling_window_length, function), "std": ...}
    scitype="tabular-regressor",
)
```

Another option would be to pass a list of transformers in make_reduction:

```python
my_reduced_regressor_estimator = make_reduction(
    regressor,
    strategy="direct",
    window_length=24,
    featurizers=[
        MovingAverageFeaturizer(10, 3), 
        ... , 
        DummyCalendarFeaturizer(week_of_year=True, month_of_year=True),
    ]
    scitype="tabular-regressor",
)
```

I believe the second option will give us more flexibility. Maybe we can also add an extension template for custom reduction featurizers to make it customisable.

Also for the recursive case, we either need to store a subset of the training set based on the maximum lag for the summary features e.g. ```self._training_y``` to allow for recomputing the summary stats every time we add the new recursive feature or pass y_train in predict (or something smarter than those 2).

**Comment by LoicRaillon** [comment]
I think we should keep in mind that `make_reduction` doesn't support exogenous variables `X` currently but if it does in the future (I hope so), we should take it into account when designing the API. 
As suggested by @ltsaprounis, we could do this by mapping a variable name to a `summary_features` or `featurizers`.

**Comment by fkiraly** [comment]
> I think we should keep in mind that make_reduction doesn't support exogenous variables X currently but if it does in the future (I hope so),

@loicraillon, I think that some reduction methods do support `X` - the direct reduction strategy does, for instance?

**Comment by fkiraly** [comment]
@ltsaprounis, I think you are right - if you get features from `y`, this probably needs to be part of the `make_reduction`.
I prefer the variant where you can pass transformers.

Summarizing `X`, however, works with a transformer, no?
More generally, a "general" pipeline would work with transformers only, if you have `x_steps`, `y_steps`, and `y_to_x_steps`, similar to what @aiwalter was proposing?

In any case, would you mind updating the issue with the specification you are implementing?

**Comment by fkiraly** [comment]
to summarize discussion we had today in our regular meeting on panel/global forecasting:
(FYI @dwarac-PCI, @loicraillon)

an earlier design we had, on lagged feature generation and lagged reduction to regression, was already suitable to provide the functionality above, in combination with the pipeline construct.

It just never got turned into a proper issue, here it is: #1612.
@danbartl, @ltsaprounis would be working on this.

@loicraillon, @dwarak-PCI, we would very much appreciate comments on #1612.

**Comment by fkiraly** [comment]
Btw, @Dwarak-PCI, @LoicRaillon, if you would like to be involved in the discussions: there's a public invite to the regular meetings in the forecasting global/panel/hierarchical work stream on the sktime calendar and slack. You can view notes, designs, and the regular meetings in the sktime slack general channel and in the forecasting-global channel.

**Comment by fkiraly** [comment]
@Dwarak-PCI, I think this is now supported as of 0.11.2: `make_reduction` can use a `transformers` argument, to which you can pass a `WindowSummarizer`.

Some combinations of usage may not support this though, I've opened https://github.com/alan-turing-institute/sktime/issues/2465 to ensure we cover all combination cases.
